src/java/org/apache/cassandra/index/sai/disk/v1/SSTableIndexWriter.java

```diff
@@ -213,18 +219,53 @@ else if (shouldFlush(sstableRowId))
             currentBuilder = newSegmentBuilder();
         }
 
-        if (term.remaining() == 0) return;
+        if (term.remaining() == 0)
+            return;
 
-        if (!TypeUtil.isLiteral(type))
+        updatesInFlight.incrementAndGet();
+        ForkJoinPool.commonPool().submit(() -> {
```

Comment:
`SegmentBuilder` is not thread safe:

```
@NotThreadSafe
public abstract class SegmentBuilder
```

Reply:
this will take some restructuring

```
        if (rowCount == 0)
        {
            // use first global rowId in the segment as segment rowId offset
            segmentRowIdOffset = sstableRowId;
        }
```